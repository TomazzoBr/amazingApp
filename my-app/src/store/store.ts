import { configureStore } from "@reduxjs/toolkit";
import toggleFilterReducer from "./reducers/toggleFilterReducer";
import toggleFavouriteModalReducer from "./reducers/toggleFavouriteModalReducer";

const reducers = {
  toggleFilterReducer,
  toggleFavouriteModalReducer,
};

const store = configureStore({ reducer: reducers });

export default store;
