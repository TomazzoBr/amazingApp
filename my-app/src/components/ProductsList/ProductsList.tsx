import { useEffect, useState } from "react";
import { ProductsInterface } from "../../interfaces/ProductsInterface";

import ProductItem from "./ProductItem.tsx/ProductItem";
import Pagination from "./Pagination/Pagination";
import SortBar from "../SortBar/SortBar";
import store from "../../store/store";

export default function Products() {
  const [productsData, setProductsData] = useState<ProductsInterface[]>([]);
  const [page, setPage] = useState(1);
  const [totalPages, setTotalPages] = useState(1);

  const handlePrevPage = (prevPage: number) => {
    setPage((prevPage) => prevPage - 1);
  };

  const handleNextPage = (nextPage: number) => {
    setPage((nextPage) => nextPage + 1);
  };

  const getPaginatedData = () => {
    const itemsPerPage = 5;
    const startIndex = page * itemsPerPage - itemsPerPage;
    const endIndex = startIndex + itemsPerPage;
    return productsData.slice(startIndex, endIndex);
  };

  useEffect(() => {
    const fetchData = async () => {
      const response = await fetch(
        `https://frontend-tech-test-data.s3.eu-west-1.amazonaws.com/items.json`
      );
      const result = await response.json();
      setProductsData(result.items);
      setTotalPages(result.items.length / 5);
    };
    fetchData();
  });

  const toggleFilterValue = store.getState().toggleFilterReducer.value;

  if (toggleFilterValue === "title") {
    productsData.sort((a, b) => {
      if (a.title < b.title) {
        return -1;
      }
      if (a.title > b.title) {
        return 1;
      }
      return 0;
    });
  } else if (toggleFilterValue === "description") {
    productsData.sort((a, b) => {
      if (a.description < b.description) {
        return -1;
      }
      if (a.description > b.description) {
        return 1;
      }
      return 0;
    });
  } else if (toggleFilterValue === "price") {
    productsData.sort((a, b): number => {
      return a.price - b.price;
    });
  } else if (toggleFilterValue === "email") {
    productsData.sort((a, b) => {
      if (a.email < b.email) {
        return -1;
      }
      if (a.email > b.email) {
        return 1;
      }
      return 0;
    });
  }

  return (
    <div className="w-full flex-wrap justify-center">
      <div className="w-full h-8 flex justify-around items-center  mt-4">
        <SortBar />
        <Pagination
          totalPages={totalPages}
          currentPage={page}
          handlePrevPage={handlePrevPage}
          handleNextPage={handleNextPage}
        />
      </div>
      <div className="w-full flex-wrap">
        <ProductItem
          getPaginatedData={getPaginatedData}
          productsData={productsData}
        />
      </div>
    </div>
  );
}
