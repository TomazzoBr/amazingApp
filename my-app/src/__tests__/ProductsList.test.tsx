import React from "react";
import renderer from "react-test-renderer";
import { cleanup } from "@testing-library/react";
import dataJSON from "../assets/data.json";

import ProductsList from "../components/ProductsList/ProductsList";
import { Provider } from "react-redux";
import store from "../store/store";

afterEach(cleanup);

describe("<ProductsList />", () => {
  const products = JSON.parse(JSON.stringify(dataJSON.items));
  test("should display the initial status of the products list", async () => {
    const productsComponent = renderer.create(
      <Provider store={store}>
        <ProductsList filteredProd={products.sort()} products={products} />
      </Provider>
    );
    expect(productsComponent).toMatchSnapshot();
  });
});
