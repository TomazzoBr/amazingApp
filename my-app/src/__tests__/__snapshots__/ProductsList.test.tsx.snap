// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductsList /> should display the initial status of the products list 1`] = `
<div
  className="w-full h-full flex flex-wrap justify-center"
>
  <div
    className="w-3/4 h-12 flex justify-around items-center bg-backgreen mt-2 rounded-md"
  >
    <div
      className="w-1/2"
    >
      <form
        className="flex flex-row justify-center items-center"
      >
        <label
          className="bg-white p-1 rounded-md shadow-sm"
          htmlFor="products-select"
        >
          Sort by:
        </label>
        <select
          className="w-1/3 h-1/3 ml-2 rounded-md shadow-md px-4 py-2 bg-white text-sm font-medium text-gray-700 hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-offset-gray-100 focus:ring-indigo-500"
          onChange={[Function]}
        >
          <option
            value="none"
          >
            No filter
          </option>
          <option
            value="title"
          >
            Title
          </option>
          <option
            value="description"
          >
            Description
          </option>
          <option
            value="price"
          >
            Price
          </option>
          <option
            value="email"
          >
            E-mail
          </option>
        </select>
      </form>
    </div>
    <div
      className="flex items-center"
    >
      <button
        className="bg-white mr-2 rounded-xl p-1 shadow-md hover:shadow-none"
        disabled={true}
        onClick={[Function]}
      >
        ←
      </button>
      <span
        className="bg-white rounded-md p-1 shadow-md"
      >
        Page 
        1
         of 
        1
      </span>
      <button
        className="bg-white ml-2 rounded-xl p-1 shadow-md hover:shadow-none"
        disabled={true}
        onClick={[Function]}
      >
        →
      </button>
    </div>
  </div>
  <div
    className="relative w-full h-full flex flex-wrap justify-center"
  >
    <div
      className="flex w-full h-full p-3"
    >
      <ul
        className="w-screen flex p-4 justify-center bg-backblue rounded-md"
      >
        <li
          className="flex items-center"
        >
          <strong>
            No products available :(
          </strong>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
